File: data_structure/1_linked_list/6_swap_nodes/swap_nodes.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:

        if not head or not head.next:
            return head

        prevNode = None

        def swapTwo(node1, node2):
            nonlocal prevNode
            if prevNode:
                prevNode.next = node2
            temp = node2.next
            node2.next = node1
            node1.next = temp
            return node1

        counter = 1

        current = head
        output = head.next

        while current and current.next:
            if counter%2 == 1:
                newCurrent = swapTwo(current,current.next)
                current = newCurrent
            else:
                prevNode = current
                current = current.next
            counter = counter + 1

        return output

File: data_structure/1_linked_list/6_swap_nodes/swap_nodes.js
'use strict';

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
  if (!head || !head.next) return head;

  let prevNode = null;
  let output = head.next;
  let counter = 1;
  let current = head;
  while(current && current.next){
    if (counter%2===1){
      current = swapTwo(current, current.next);
    } else {
      prevNode = current;
      current = current.next;
    }
    counter++;
  }
  return output;

  function swapTwo(node1, node2){
    if (prevNode) prevNode.next = node2;

    let temp = node2.next;
    node2.next = node1;
    node1.next = temp;

    return node1;
  }
};

module.exports = swapPairs;
